src/analysis/souffle/taint_with_auth_logic.dl

```diff
@@ -0,0 +1,114 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+
+// This file gives rules that define a check that does information flow control,
+// but uses authorization logic for downgrading.
+
+#include "flat_graph_ir.dl"
+#include "auth_logic_interface.dl"
+
+#ifndef TAINT_WITH_AUTH_LOGIC
+#define TAINT_WITH_AUTH_LOGIC
+
+//-----------------------------------------------------------------------------
+// Type Declarations
+//-----------------------------------------------------------------------------
+// A taint Tag, such as "video_data"
+.type Tag <: symbol
+
+// An information flow Label which in this model is a set of taint Tags
+.type Label <: symbol
+
+//-----------------------------------------------------------------------------
+// Predicate Declarations
+//-----------------------------------------------------------------------------
+// There is a claim that `accessPath` definitely has `tag` taint.
+.decl claimHasTag(accessPath: AccessPath, tag: Tag)
+
+// The Label of a accesPath has a Tag according to the taint analysis
+// NOTE: In practice there may be value in explicitly having a layer of
+// indirection in which accessPaths have Labels and just Labels have Tags.
+// In this formalization, accessPaths also act as Labels (in the sense that they
+// have Tags).
+.decl mayHaveTag(p: AccessPath, t: Tag)
+
+// This party is the owner of a taint Tag (they have the privilege to downgrade
+// it).
+.decl ownsTag(owner: Principal, t: Tag)
+
+// Some accessPath is actually downgraded (by taking into consideration the 
+// privileges of a party trying to downgrade it)
+.decl downgrades(p: AccessPath, t: Tag)
+
+// A Tag is a member of the Label ell (which is a set of Tags).
+.decl isMember(t: Tag, ell: Label)
+
+// The Label of an accessPath does not flowTo another Label
+// This is equivalent to "not (subset or equals)"
+.decl notFieldLabelFlowsTo(p: AccessPath, ell: Label)
```

Comment:
Another optional suggestion: I find that it can be confusing to maintain code that contains too many negatives. It might be clearer to define a relation `fieldLabelFlowsTo` and to define `notFieldLabelFlowsTo` in terms of that.  If you need `notFieldLabelFlowsTo` at all, that is: see comment later in file.